core/internal/runfiles/filesrecordhandler.go

```diff
@@ -0,0 +1,98 @@
+package runfiles
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/wandb/wandb/core/internal/filetransfer"
+	"github.com/wandb/wandb/core/internal/settings"
+	"github.com/wandb/wandb/core/pkg/observability"
+	"github.com/wandb/wandb/core/pkg/service"
+)
+
+// Internal implementation of the FilesRecordHandler interface.
+type handler struct {
+	logger       *observability.CoreLogger
+	settings     *settings.Settings
+	fileTransfer filetransfer.FileTransferManager
+	graphQL      graphql.Client
+
+	// Wait group for file uploads.
+	uploadWg *sync.WaitGroup
+
+	// Set of files (not directories) to upload later.
+	flushSet map[string]FileInfo
+
+	// Whether 'Finish' was called.
+	isFinished *atomic.Bool
+
+	// Mutex that's locked whenever any state is being read or modified.
+	stateMu *sync.Mutex
+}
+
+func newFilesRecordHandler(params FilesRecordHandlerParams) FilesRecordHandler {
+	return &handler{
+		logger:       params.Logger,
+		settings:     params.Settings,
+		fileTransfer: params.FileTransfer,
+		graphQL:      params.GraphQL,
+
+		uploadWg: &sync.WaitGroup{},
+
+		flushSet: make(map[string]FileInfo),
+
+		isFinished: &atomic.Bool{},
+
+		stateMu: &sync.Mutex{},
+	}
+}
+
+func (h *handler) ProcessRecord(record *service.FilesRecord) {
+	if h.isFinished.Load() {
+		h.logger.CaptureError("runfiles: called ProcessRecord() after Finish()", nil)
+		return
+	}
+
+	h.stateMu.Lock()
+	defer h.stateMu.Unlock()
+	// TODO
+}
+
+func (h *handler) Flush() {
+	if h.isFinished.Load() {
+		h.logger.CaptureError("runfiles: called Flush() after Finish()", nil)
+		return
+	}
+
+	h.stateMu.Lock()
+	defer h.stateMu.Unlock()
+
+	for path, info := range h.flushSet {
+		// Avoid capturing the loop variables.
```

Comment:
i think we go 1.22 fixed it: https://go.dev/blog/loopvar-preview

Reply:
and here : https://go.dev/blog/go1.22 :)

Reply:
wow! glad they fixed it!

Go wants to stay on major version 1 so badly that they make breaking changes in minor versions :p

Reply:
unfortunately it fails pre-push checks:

<img width="1012" alt="Screenshot 2024-03-19 at 6 39 03 PM" src="https://github.com/wandb/wandb/assets/6894354/71419ab4-454d-43ec-8f1f-93ce0af2e817">

i guess we need to wait for the linters to update

Reply:
oh wow... interesting, do we have the latest version of the tool?

Reply:
we probably need to bump a version in some CI config somewhere

Reply:
i removed this code from the PR though

Reply:
yeah, i meant to update our sum file so i'll see if we can update all our tools as well